src/mygrad/nnet/losses/focal_loss.py

```diff
@@ -50,54 +51,70 @@ def __call__(self, scores, targets, alpha, gamma):
         numpy.ndarray
             The per-datum focal loss.
         """
-        if isinstance(targets, Tensor):
+        if isinstance(targets, Tensor):  # pragma: nocover
             targets = targets.data
 
-        check_loss_inputs(scores, targets)
+        check_loss_inputs(class_probs, targets)
 
-        self.variables = (scores,)
-        scores = np.copy(scores.data if isinstance(scores, Tensor) else scores)
+        self.variables = (class_probs,)
+        if isinstance(class_probs, Tensor):
+            class_probs = class_probs.data
 
-        max_scores = np.max(scores, axis=1, keepdims=True)
-        np.exp(scores - max_scores, out=scores)
-        scores /= np.sum(scores, axis=1, keepdims=True)
-        label_locs = (range(len(scores)), targets)
+        self.label_locs = (range(len(class_probs)), targets)
 
-        pc = scores[label_locs]
-        one_m_pc = 1 - pc + 1e-14  # correct domain for when gamma < 1 and pc == 1
+        pc = class_probs[self.label_locs]
+        one_m_pc = np.clip(1 - pc, a_min=0, a_max=1)
         log_pc = np.log(pc)
 
-        loss = -(alpha * one_m_pc ** gamma * log_pc)
-
-        self.back = scores
-        self.back[label_locs] -= 1
-        deriv = one_m_pc ** gamma - pc * gamma * one_m_pc ** (gamma - 1) * log_pc
-        self.back *= deriv[:, np.newaxis]
-        self.back *= alpha
-        return loss
+        one_m_pc_gamma = one_m_pc ** gamma
+        loss = -(alpha * one_m_pc_gamma * log_pc)
+
+        self.back = np.zeros(class_probs.shape, dtype=np.float64)
+
+        if np.isclose(gamma, 0, atol=1e-15):
+            self.back[self.label_locs] -= alpha / pc
+            return loss
+        else:
+            # dL/dp = (1 - p)**g / p - g * (1 - p)**(g - 1) * log(p)
+            term1 = one_m_pc_gamma / pc  # (1 - p)**g / p
+            if np.isclose(gamma, 1, rtol=1e-15):
+                term2 = -log_pc
+            elif gamma < 1:
+                # For g < 1 and p -> 1, the 2nd term -> 0 via L'Hôpital's rule
+                term2 = np.zeros(pc.shape, dtype=np.float64)
+                pc_not_1 = ~np.isclose(one_m_pc, 0, atol=1e-15)
+                term2[pc_not_1] = (
+                    -gamma * one_m_pc[pc_not_1] ** (gamma - 1) * log_pc[pc_not_1]
+                )
+            else:
+                term2 = -gamma * one_m_pc ** (gamma - 1) * log_pc
+
+            self.back[self.label_locs] -= alpha * (term1 + term2)
+            return loss
```

Comment:
```suggestion
        return loss
```
to account for the above